### ISSUE TITLE
Workflow missing volume in spec when generated from workflow template that has volume defined

### ISSUE THREAD

**Issue by jmeridth** [argoproj/argo-workflows#7606]
<!--
Before we start, around 2/3 of issues can be fixed by one of the following:

* Have you double-checked your configuration? Maybe 30% of issues are wrong configuration.
* Have you tested to see if it is fixed in the latest version? Maybe 20% of issues are fixed by this.
* Have you tried using the PNS executor instead of Docker? Maybe 50% of artifact related issues are fixed by this.

If this is a regression, please open a regression report instead.
-->

## Summary

Relates to https://github.com/argoproj/argo-workflows/issues/3156#issuecomment-790915195

Generate workflow from workflow template that has a volume & volume mount defined.  The workflow has the volume mount but not the volume.

Expected:
Workflow succeeds because volume in workflow spec

Actual:
Error: `task 'webhook-f4hmg[0].template1-step.step1' errored: volume 'blah-blah-blah' not found in workflow spec`

Argo Workflow Version: v3.2.6

##

Example Yaml (test.yaml)

```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: hello-world-template-global-arg
spec:
  serviceAccountName: argo
  volumes:
  - name: ssh-git-creds
    secret:
      defaultMode: 0600
      secretName: ssh-git-creds
      items:
        - key: ssh_privatekey
          path: id_rsa
        - key: known_hosts
          path: known_hosts
  templates:
    - name: hello-world
      container:
        image: docker/whalesay
        command: [cowsay]
        args: ["{{workflow.parameters.global-parameter}}"]
        volumeMounts:
        - name: ssh-git-creds
          mountPath: /root/.ssh
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: hello-world-wf-global-arg-1
spec:
  serviceAccountName: argo
  entrypoint: whalesay
  arguments:
    parameters:
      - name: global-parameter
        value: hello
  templates:
    - name: whalesay
      steps:
        - - name: hello-world
            templateRef:
              name: hello-world-template-global-arg
              template: hello-world

```

## Diagnostics

Deploy

```bash
kubectl apply -f test.yaml
workflowtemplate.argoproj.io/hello-world-template-global-arg created
workflow.argoproj.io/hello-world-wf-global-arg-1 created
```

What Kubernetes provider are you using? Minikube/K8s 1.22

What executor are you running? Docker

```bash
➜  ~ k get workflows                                                                                                                                                                                                          13:28:06
NAME                          STATUS      AGE
hello-world-wf-global-arg-1   Failed      5s
```

```bash
 k get workflow hello-world-wf-global-arg-1 -o yaml
```

partial result

```bash

... (NOTE THE message about 'volume ... not found in workflow spec')
nodes:
    hello-world-wf-global-arg-1:
      children:
      - hello-world-wf-global-arg-1-1009745991
      displayName: hello-world-wf-global-arg-1
      finishedAt: "2022-01-20T19:28:07Z"
      id: hello-world-wf-global-arg-1
      message: 'step group deemed errored due to child hello-world-wf-global-arg-1[0].hello-world
        error: volume ''ssh-git-creds'' not found in workflow spec'
      name: hello-world-wf-global-arg-1
      outboundNodes:
      - hello-world-wf-global-arg-1-3687095688
      phase: Failed
      progress: 1/1
      startedAt: "2022-01-20T19:28:07Z"
      templateName: whalesay
      templateScope: local/hello-world-wf-global-arg-1
      type: Steps
...
```

---
<!-- Issue Author: Don't delete this message to encourage other users to support your issue! -->
**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by jmeridth** [comment]
Determined this was an argo-workflows situation from the issue mentioned above and also after working through our problem in [this argo-events issue](https://github.com/argoproj/argo-events/issues/1542) (same issue but thought it might be argo-events and is not) with @whynowy.

**Comment by jmeridth** [comment]
Is `templateDefault` the only current solution like mentioned [here](https://github.com/argoproj/argo-workflows/issues/3156#issuecomment-923434425)?

**Comment by pisymbol** [comment]
This is also causing k8s based executors to barf with the emptyDir() error when trying to use parameterized outputs. It looks like the volume spec isn't being evaluated if defined in templateDefault before the check.

**Comment by jmeridth** [comment]
`templateDefault` and `workflowDefault` defined volumes are not being add to the generated workflow spec either for me.  Kind of curious why a generated workflow from a workflow template in the dashboard or argo cli keep the volume in the workflow spec, but if used with `kubectl` == no joy. 🤔

**Comment by alexec** [comment]
This is because you're using `templateRef` not `workflowTemplateRef`. `templateRef` does not inherit volumes (or anything else) from the `spec`.

**Comment by pisymbol** [comment]
@alexec  Can you explain why templateDefaults do not include volumeMounts?

**Comment by alexec** [comment]
This issue is not about `templateDefaults` AFAIK. Maybe raise another issue?

**Comment by jmeridth** [comment]
@alexec confirmed.  We needed to use `workflowTemplateRef` instead of `templateRef`.  Thank you.

For others, solution is:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: hello-world-template-global-arg
spec:
  serviceAccountName: argo
  volumes:
  - name: ssh-git-creds
    secret:
      defaultMode: 0600
      secretName: ssh-git-creds
      items:
        - key: ssh_privatekey
          path: id_rsa
        - key: known_hosts
          path: known_hosts
  templates:
    - name: hello-world
      container:
        image: docker/whalesay
        command: [cowsay]
        args: ["{{workflow.parameters.global-parameter}}"]
        volumeMounts:
        - name: ssh-git-creds
          mountPath: /root/.ssh
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: hello-world-wf-global-arg-1
spec:
  serviceAccountName: argo
  entrypoint: hello-world
  arguments:
    parameters:
      - name: global-parameter
        value: hello
  workflowTemplateRef:
    name: hello-world-template-global-arg
```

change `entrypoint` to entry point on workflow template and use `workflowTemplateRef` right under arguments.  Example in docs [here](https://github.com/argoproj/argo-workflows/blob/master/examples/workflow-template/workflow-template-ref-with-entrypoint-arg-passing.yaml)

@alexec did you want me to close this issue?  I saw you tagged it for `enhancement`.  Don't know what plans you might have.  Cheers.

**Comment by alexec** [comment]
I think so, I did think maybe we should copy the volumes, but I think we should not do that.

**Comment by mxm-tr** [comment]
I've also reproduced the issue but using a DAGTask, and I haven't been able to use the workflowTemplateRef there.

According to the [reference](https://argoproj.github.io/argo-workflows/fields/#dagtask) only templateRef seems to be available in a DAGTask

**Comment by ccureau** [comment]
I'm trying to get this same thing to work with multiple Workflow Templates that I want to run from a specific template. It seems that a `workflowTemplateRef` can only reference one workflow template.

**Comment by jmeridth** [comment]
Closing this issue.  If someone want to reopen and create PR to get what we need from this, please do so.  I'd suggest a new issue and link to this for reference.  Cleaner that way.  Cheers.

**Comment by Mantower** [comment]
Did anyone ever manage to setup a WorkflowTemplate in the shape of a DAG that references other workflow templates with volumes? Our current setup utilises a template like this:

```
# Source: mlops-helm/templates/workflows.yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: master-template
spec:
  serviceAccountName: pod-sa
  entrypoint: sequence
  podGC: 
    strategy: OnWorkflowSuccess
  volumes:
  - name: data
    emptyDir:
      sizeLimit: 10Gi
  templates:
  - name: sequence
    dag:
      tasks:
        - name: template-2
          dependencies:
            - template-1
          templateRef:
            name: workflowtemplate-1
            template: template-2
        - name: template-3
          dependencies:
            - template-2
          templateRef:
            name: workflowtemplate-2
            template: template-1
        - name: template-1
          templateRef:
            name: workflowtemplate-1
            template: template-1
```

where the templates in `workflowtemplate-1` looks like this:


```
# Source: mlops-helm/templates/workflows.yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: workflowtemplate-1
spec:
  serviceAccountName: pod-sa
  entrypoint: template-1
  podGC: 
    strategy: OnWorkflowSuccess
  volumes:
  - name: data
    emptyDir:
      sizeLimit: 10Gi
  templates:
  - name: template-2
    container:
      image: <image>
      volumeMounts:
        - containerPath: "/data"
          volume:
            name: data
            readOnly: false
  - name: template-1
    container:
      image: <image>
      volumeMounts:
        - containerPath: "/data"
          volume:
            name: data
            readOnly: false
```
